src/jit/morph.cpp

```diff
@@ -4984,8 +4984,9 @@ GenTreePtr Compiler::fgMorphMultiregStructArg(GenTreePtr arg, fgArgTabEntryPtr f
             }
         }
     }
-    // We should still have a TYP_STRUCT
-    assert(varTypeIsStruct(argValue->TypeGet()));
+    // We should still have a TYP_STRUCT or
+    // if struct is produced from `localloc` IL code we might have TYP_BLK
+    assert(varTypeIsStruct(argValue->TypeGet()) || (argValue->TypeGet() == TYP_BLK));
```

Comment:
What do we have here? An outgoing argument with a TYP_BLK produced from a localloc?
And we are trying to pass it as a multi-reg Struct?
Where did this IL come from?

Reply:
Yes, you are right.

TYP_BLK was initialized into value type by `initobj` 

It is from `JIT\Methodical\tailcall\test_3b.il`

Reply:
The new behavior is likely a consequence of #14623.

Reply:
@AndyAyersMS very likely. The code creates such locals is importer.cpp : 14355 in the mentioned PR

Reply:
How a struct is to be passed  often requires knowledge about the struct's fields.
We might have to pass  this in floating point registers.  Have a TYP_BLK here can be a problem.

I think that you just need add new code here to special case the TYP_BLK case here entirely.

Reply:
@briansull If I am correct any transformation is based on root type which is `TYP_STRUCT` regardless what exactly holds it (`TYP_BLK` or whatever else). Could you please clarify how should we handle `TYP_BLK`?

Reply:
The passing of multi reg args is complex.  

We only handle three case currently

argValue == GT_LCL_VAR or GT_LCL_FLD  or GT_OBJ

Even for these there are slightly different cases for each of these:
1. MultiReg sized Struct arg that is passed in register
       Is a struct promoted argument used as the actual argument value
         Form a N-element FIELD_LIST to hold each of the register arguments
2. Multireg sized Struct arg that is passed on the stack (because we used up all register arguments)

When moving value around keep track of the GC layout so we know if we are copying a gc pointer or a byref pointer.  (We should check that we are handling Span correctly here, we might not be)

A TYP_BLK case would be mostly similar to the TYP_OBJ case and I don't think that it can contain any GC pointers.

Reply:
@briansull I am still a bit confused. Are you sure about `TYP_OBJ`? It doesn't seem that we have such type. I mean `TYP_BLK` is node's type, not a kind of operator as `GT_LCL_VAR` is.

Reply:
The point that @briansull is making is that `TYP_BLK` is not an expected type for an argument. For this case the node should be re-typed as `TYP_STRUCT`, and the `objClass` should be obtained from the argument signature, if we are passing a local that was produced from a localloc and we are passing an effectively untyped block.
Even if there are no GC fields, for some architectures, we will still need to know the layout in order to determine how it is passed, e.g. possibly using some floating point registers.